{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualization & Analysis of Chemical Data\n",
    "This week we will look at a few ways to visualize and analyze chemical data. We already learned some ways to analyze chemical data in weeks 5 and 6 when we got familiar with RDKit. Visualizing through smart ways of plotting data is also a very important part of data analysis, which allows you to get a feeling of your data and to identify patterns.\n",
    "\n",
    "Frst, let's get some housekeeping out of the way. We will use the same data as in weeks 5 and 6 from ChEMBL and filter it right away by excluding data with missing SMILES and USAN Definitions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some imports\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import PandasTools\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent Molecule</th>\n",
       "      <th>Name</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Research Codes</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Drug Applicants</th>\n",
       "      <th>USAN Stem</th>\n",
       "      <th>USAN Year</th>\n",
       "      <th>USAN Definition</th>\n",
       "      <th>USAN Stem - Substem</th>\n",
       "      <th>...</th>\n",
       "      <th>Passes Rule of Five</th>\n",
       "      <th>First In Class</th>\n",
       "      <th>Chirality</th>\n",
       "      <th>Prodrug</th>\n",
       "      <th>Oral</th>\n",
       "      <th>Parenteral</th>\n",
       "      <th>Topical</th>\n",
       "      <th>Black Box</th>\n",
       "      <th>Availability Type</th>\n",
       "      <th>Smiles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL492491</td>\n",
       "      <td>FENVALERATE</td>\n",
       "      <td>BELMARK|FENVALERATE|PHENVALERATE|PYDRIN|S-5602...</td>\n",
       "      <td>S-5602|SD-43775|WL-43775</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Racemic Mixture</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CC(C)C(C(=O)OC(C#N)c1cccc(Oc2ccccc2)c1)c1ccc(C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL2103777</td>\n",
       "      <td>THIOPHANATE</td>\n",
       "      <td>NSC-170810|THIOPHANATE</td>\n",
       "      <td>NSC-170810</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Achiral Molecule</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CCOC(=O)NC(=S)Nc1ccccc1NC(=S)NC(=O)OCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1868702</td>\n",
       "      <td>GESTRINONE</td>\n",
       "      <td>A 46 745|A-46-745|A-46745|DIMETRIOSE|GESTRINON...</td>\n",
       "      <td>A 46 745|A-46-745|A-46745|R 2323|R-2323|RU 232...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>'-estr-; -rinone'</td>\n",
       "      <td>1978.0</td>\n",
       "      <td>estrogens; cardiotonics (amrinone type)</td>\n",
       "      <td>'-estr-(-estr-); -rinone(-rinone)'</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Single Stereoisomer</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>C#C[C@]1(O)CC[C@H]2[C@@H]3CCC4=CC(=O)CCC4=C3C=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL2104213</td>\n",
       "      <td>DESMENINOL</td>\n",
       "      <td>DESMENINOL|METHIONINE HYDROXY ANALOG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Racemic Mixture</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CSCCC(O)C(=O)O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL1904952</td>\n",
       "      <td>QUAZODINE</td>\n",
       "      <td>MJ 1988|MJ-1988|QUAZODINE</td>\n",
       "      <td>MJ 1988|MJ-1988</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1968.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Achiral Molecule</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CCc1ncnc2cc(OC)c(OC)cc12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Parent Molecule         Name  \\\n",
       "0    CHEMBL492491  FENVALERATE   \n",
       "1   CHEMBL2103777  THIOPHANATE   \n",
       "2   CHEMBL1868702   GESTRINONE   \n",
       "3   CHEMBL2104213   DESMENINOL   \n",
       "4   CHEMBL1904952    QUAZODINE   \n",
       "\n",
       "                                            Synonyms  \\\n",
       "0  BELMARK|FENVALERATE|PHENVALERATE|PYDRIN|S-5602...   \n",
       "1                             NSC-170810|THIOPHANATE   \n",
       "2  A 46 745|A-46-745|A-46745|DIMETRIOSE|GESTRINON...   \n",
       "3               DESMENINOL|METHIONINE HYDROXY ANALOG   \n",
       "4                          MJ 1988|MJ-1988|QUAZODINE   \n",
       "\n",
       "                                      Research Codes  Phase Drug Applicants  \\\n",
       "0                           S-5602|SD-43775|WL-43775   -1.0             NaN   \n",
       "1                                         NSC-170810   -1.0             NaN   \n",
       "2  A 46 745|A-46-745|A-46745|R 2323|R-2323|RU 232...    4.0             NaN   \n",
       "3                                                NaN   -1.0             NaN   \n",
       "4                                    MJ 1988|MJ-1988   -1.0             NaN   \n",
       "\n",
       "           USAN Stem  USAN Year                          USAN Definition  \\\n",
       "0                NaN        NaN                                      NaN   \n",
       "1                NaN        NaN                                      NaN   \n",
       "2  '-estr-; -rinone'     1978.0  estrogens; cardiotonics (amrinone type)   \n",
       "3                NaN        NaN                                      NaN   \n",
       "4                NaN     1968.0                                      NaN   \n",
       "\n",
       "                  USAN Stem - Substem  ...  Passes Rule of Five  \\\n",
       "0                                 NaN  ...                    0   \n",
       "1                                 NaN  ...                    1   \n",
       "2  '-estr-(-estr-); -rinone(-rinone)'  ...                    1   \n",
       "3                                 NaN  ...                    1   \n",
       "4                                 NaN  ...                    1   \n",
       "\n",
       "  First In Class            Chirality Prodrug Oral Parenteral Topical  \\\n",
       "0              0      Racemic Mixture       0    0          0       0   \n",
       "1              0     Achiral Molecule       0    0          0       0   \n",
       "2              0  Single Stereoisomer       0    0          0       0   \n",
       "3              0      Racemic Mixture       0    0          0       0   \n",
       "4              0     Achiral Molecule       0    0          0       0   \n",
       "\n",
       "  Black Box Availability Type  \\\n",
       "0         0           Unknown   \n",
       "1         0           Unknown   \n",
       "2         0           Unknown   \n",
       "3         0           Unknown   \n",
       "4         0           Unknown   \n",
       "\n",
       "                                              Smiles  \n",
       "0  CC(C)C(C(=O)OC(C#N)c1cccc(Oc2ccccc2)c1)c1ccc(C...  \n",
       "1             CCOC(=O)NC(=S)Nc1ccccc1NC(=S)NC(=O)OCC  \n",
       "2  C#C[C@]1(O)CC[C@H]2[C@@H]3CCC4=CC(=O)CCC4=C3C=...  \n",
       "3                                     CSCCC(O)C(=O)O  \n",
       "4                           CCc1ncnc2cc(OC)c(OC)cc12  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "current_file = Path(os.path.abspath(''))\n",
    "csv_file = current_file.parent / \"week_10\" / \"data.csv\"\n",
    "df = pd.read_csv(csv_file, sep= \";\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you get the following error in the next cell:\n",
    "```\n",
    "AttributeError: module 'pandas.io.formats.format' has no attribute 'get_adjustment'\n",
    "```\n",
    "Install pandas version 2.1 with the following command:\n",
    "```\n",
    "!pip install pandas==2.1\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the rows with NaN values for SMILES and USAN Definition\n",
    "df = df.dropna(subset=[\"Smiles\", \"USAN Definition\"])\n",
    "# add molecule objects to the df from SMILES\n",
    "PandasTools.AddMoleculeColumnToFrame(df, smilesCol=\"Smiles\", molCol=\"mol\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['estrogens; cardiotonics (amrinone type)', 'sympathomimetics',\n",
       "       'alpha-adrenoceptor antagonists (benzodioxane derivatives)',\n",
       "       'fibrinogen receptor antagonists (glycoprotein IIb/IIIa receptor antagonists)',\n",
       "       'antineoplastics (chloroethylamine derivatives)', 'antiandrogens',\n",
       "       'endopeptidase inhibitors',\n",
       "       'quaternary ammonium derivatives; quaternary ammonium derivatives',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'quaternary ammonium derivatives', 'antidiabetics', 'estrogens',\n",
       "       'prostaglandins', 'antivirals',\n",
       "       'antihyperlipidemics (clofibrate type)',\n",
       "       '5-nitrofuran derivatives',\n",
       "       'coronary vasodilators (verapamil type)',\n",
       "       'iodine-containing contrast media',\n",
       "       'anti-inflammatory/analgesic agents (ibuprofen type)',\n",
       "       'cholinesterase inhibitors (physostigmine type)',\n",
       "       'bronchodilators (phenethylamine derivatives)',\n",
       "       'anti-inflammatory agents (acetic acid derivatives)',\n",
       "       'systemic antifungals (miconazole type)',\n",
       "       'spasmolytic agents (papaverine type)',\n",
       "       'anti-inflammatory agents (indomethacin type)',\n",
       "       'analgesics (mixed opiate receptor agonists/antagonists)',\n",
       "       'antiallergics (cromoglicic acid derivatives)',\n",
       "       'antibacterial antibiotics, chloramphenicol analogs',\n",
       "       'anti-inflammatory analgesics (phenylbutazone type)',\n",
       "       'inhibitors of plasminogen activator inhibitors- type 1',\n",
       "       'phosphoro-derivatives', 'tricyclic compounds',\n",
       "       'antifungal antibiotics (undefined group)',\n",
       "       'cortisone derivatives; steroids (not prednisolone derivatives)',\n",
       "       'steroids (androgens, anabolics); antiandrogens',\n",
       "       'dopaminergics (butopamine type)',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines): type IV phosphodiesterase',\n",
       "       'antihistamines/local vasoconstrictors (antazoline type)',\n",
       "       'local anesthetics', 'antianginals (sydnone derivatives)',\n",
       "       'potassium channel agonists', 'antipsychotics (haloperidol type)',\n",
       "       'antibiotics (Streptomyces strain)',\n",
       "       'antihistaminics (histamine-H1 receptor antagonists)',\n",
       "       'antiepileptics (hydantoin derivatives)',\n",
       "       'barbituric acid derivatives',\n",
       "       '5-HT1A receptor agonists/antagonists acting primarily as neuroprotectors',\n",
       "       'acaricide/tick repellent', 'antiarrhythmics',\n",
       "       'antiprotozoal substances (metronidazole type)',\n",
       "       'antidepressants (imipramine type)',\n",
       "       'antibiotics (Micromonospora strains)',\n",
       "       'H2-receptor antagonists (cimetidine type)',\n",
       "       'steroids (not prednisolone derivatives)', 'immunosuppressives',\n",
       "       'topical steroids (acetal derivatives)',\n",
       "       'antithrombotics, blood coagulation factor XA inhibitors',\n",
       "       'antihypertensives (guanidine deriviatives)',\n",
       "       'phenylpyridine vasodilators; tricyclic compounds',\n",
       "       'angiotensin II receptor antagonists',\n",
       "       'benzodiazepine receptor agonists/antagonists',\n",
       "       'norepinephrine, serotonin, dopamine, reuptake inhibitor',\n",
       "       'thrombin inhibitors (argatroban type)',\n",
       "       'tricyclic compounds; antipsychotics (dibenzothiazepine derivatives)',\n",
       "       'antihypertensives (ACE inhibitors): antihypertensives (ACE inhibitors), diacid analogs',\n",
       "       'enzyme inhibitors', 'atropine derivatives',\n",
       "       'antimicrobials (sulfonamides derivatives)',\n",
       "       'diagnostic aids ((phenylcarbamoyl) methyl iminodiacetic acid derivatives)',\n",
       "       'Fc fusion protein',\n",
       "       'diuretics (sulfamoylbenzoic acid derivatives)',\n",
       "       'Fc fusion protein; thrombin inhibitors (argatroban type)',\n",
       "       'positive inotropic agents (pimobendan type)',\n",
       "       'N-methyl-D-asparate (NMDA) receptor antagonist/agonists, glycine recognition site',\n",
       "       'antipsychotics (dopamine D2 antagonists)',\n",
       "       'sodium channel blockers, signal transduction modulators',\n",
       "       'steroids (not prednisolone derivatives); prednisone and prednisolone derivatives',\n",
       "       'narcotic antagonists/agonists (6,7-benzomorphan derivatives)',\n",
       "       'antivirals: antivirals (acyclovir type)',\n",
       "       'ergot alkaloid deriviatives',\n",
       "       'hypnotics/tranquilizers (zopiclone type)',\n",
       "       'analgesics (meperidine type)', 'endothelin receptor antagonists',\n",
       "       'sulpride derivatives and analogs',\n",
       "       'cephalosporins; phosphoro-derivatives',\n",
       "       'antiulcer agents (benzimidazole derivatives)',\n",
       "       'androgens; sympathomimetics', 't-RNA synthetase inhibitors',\n",
       "       'cephalosporins', 'antianxiety agents (diazepam type)',\n",
       "       'antibiotics (rifamycin derivatives)',\n",
       "       'anabolic steroids; steroids (not prednisolone derivatives)',\n",
       "       'penicillins',\n",
       "       'antiarrhythmic/antianginal/antihypertensive agents, phthalazine like structure',\n",
       "       'combined alpha and beta blockers; narcotic agonists/antagonists (normorphine type)',\n",
       "       'phosphodiesterase inhibitor',\n",
       "       'steroids (androgens, anabolics); bronchodilators (phenethylamine derivatives)',\n",
       "       'antineoplastics (thiotepa derivatives)',\n",
       "       \"antianxiety agents/neuroleptics (4'-fluoro-4-piperidinobutyrophenone derivatives)\",\n",
       "       'anti-inflammatory agents (acetic acid derivatives): anti-inflammatory; pyrazole acetic acid derivatives',\n",
       "       'peptides: inhibition of growth hormone release', 'progestins',\n",
       "       'beta-blockers (propranolol type)',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives); spasmolytic agents (papaverine type)',\n",
       "       'theophylline derivatives', 'antibiotics (Aspergillus strains)',\n",
       "       'monoamine oxidase inhibitors (hydrazine derivatives); fluoroquinolone derivatives, nonantibacterial indication (e.g., antineoplastic antibiotics)',\n",
       "       'narcotic agonists/antagonists (normorphine type)',\n",
       "       'antiarrhythmics (propafenone derivatives)',\n",
       "       'anabolic steroids; antiarrhythmic/antianginal/antihypertensive agents, phthalazine like structure',\n",
       "       'antibacterials (quinolone derivatives)',\n",
       "       'nucleoside antiviral or antineoplastic agents, cytarabine or azarabine derivatives',\n",
       "       'non-benzodiazepine anxiolytics, sedatives, hypnotics',\n",
       "       'antihyperlipidemic, thyromimetic derivatives',\n",
       "       'antineoplastic antibiotics (daunorubicin type)',\n",
       "       'antiasthmatics (bradykinin antagonists)',\n",
       "       'enzyme inhibitors: aldose-reductase inhibitors',\n",
       "       'antihypertensives (ACE inhibitors)', 'peptides',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines)',\n",
       "       'combined alpha and beta blockers', 'phospholipase A2 inhibitors',\n",
       "       'antivirals; antineoplastics (uridine derivatives)',\n",
       "       'immunomodulators', 'pentamidine derivatives/analogues',\n",
       "       'gabamimetics', 'potassium channel antagonists', 'vinca alkaloids',\n",
       "       'antiestrogens of the clomifene and tamoxifen groups',\n",
       "       'steroids (not prednisolone derivatives); topical steroids (acetal derivatives)',\n",
       "       'vasodilators (undefined group)', 'anorexiants',\n",
       "       'steroids (androgens, anabolics); antiandrogens; vinca alkaloids',\n",
       "       'analgesics (mixed opiate receptor agonists/antagonists); estrogens',\n",
       "       'benzodiazepine receptor agonists/antagonists; cannabinol derivatives',\n",
       "       'natural antibiotics (undefined group)',\n",
       "       'ionotropic non-NMDA (N-methyl D-aspartate) glutamate receptors [AMPA (alpha-amino-3-hydroxy-5-methyl-4-isoxazolepropionic acid receptor) and/or KA (kainite antagonist) receptors]: antagonists',\n",
       "       'antagonists of GABA (gamma-aminobutyric acid) regulated chloride channels, antiparasitic agents',\n",
       "       'serotonin 5-HT3 receptor antagonists',\n",
       "       'tricyclic compounds; antivirals', 'exportin 1 inhibitors',\n",
       "       'platelet aggregation inhibitors, primarily platelet P2Y12 receptor antagonists',\n",
       "       'anabolic steroids',\n",
       "       'quaternary ammonium derivatives: neuromuscular blocking agents; quaternary ammonium derivatives: neuromuscular blocking agents',\n",
       "       'antineoplastics (mitotic inhibitors; tubulin binders)',\n",
       "       'antibacterials (Actinoplanes strains)',\n",
       "       'serotonin 5-HT6 receptor inhibitors',\n",
       "       'antidepressants (dibenzol[a,d]cycloheptane derivatives)',\n",
       "       'ribofuranil derivatives (pyrazofurin type)',\n",
       "       'enzyme inhibitors: antihyperlipidemics (HMG-CoA inhibitors)',\n",
       "       'antihypertensives (hydrazine-phthalazines); antiarrhythmic/antianginal/antihypertensive agents, phthalazine like structure',\n",
       "       'amfetamine derivatives', 'cannabinol derivatives',\n",
       "       'enzyme inhibitors: elastase inhibitors',\n",
       "       'peroxisome proliferator activiating receptor (PPAR) agonists (not thiazolidene derivatives)',\n",
       "       'carbonic anhydrase inhibitors',\n",
       "       'iodine-containing contrast media; uracil type antineoplastics; uracil derivatives used as thyroid antagonists and as antineoplastics',\n",
       "       'antibacterials, pleuromulin derivatives',\n",
       "       'steroids (androgens, anabolics); testosterone reductase inhibitors',\n",
       "       'narcotic antagonists/agonists (morphinan derivatives)',\n",
       "       'acridine derivatives', 'retinol derivatives',\n",
       "       'calcium channel blockers (diltiazem type)',\n",
       "       'antimigraine agents (5-HT1 receptor agonists); sumatriptan derivatives',\n",
       "       'muscarinic receptor agonists/antagonists; quaternary ammonium derivatives: muscarinic receptor agonists/antagonists',\n",
       "       'cyclooxygenase-2 inhibitors',\n",
       "       'antineoplastics (camptothecin derivatives)',\n",
       "       'antineoplastics, alkylating agents (methane sulfonate derivatives)',\n",
       "       'cholecystokinin receptor antagonists',\n",
       "       'peptides: synthetic corticotropins',\n",
       "       'nootropic agents (learning, cognitive enhancers), piracetam type',\n",
       "       'anxiolytics (buspirone type)', 'sphingosine derivatives',\n",
       "       'enzyme inhibitors: aldose-reductase inhibitors; enzyme inhibitors: urease inhibitors',\n",
       "       'beta3 adrenoreceptor agonists', 'sigma receptor ligands',\n",
       "       'antipsychotics (risperidone type)', 'glucokinase activator',\n",
       "       'serine/threonine kinase inhibitors',\n",
       "       'steriodal compounds acting on progesterone receptors (excluding-gest- compounds); progesterone receptor antagonists',\n",
       "       'antibiotics (tetracycline derivatives)',\n",
       "       'analgesics (fenamic acid subgroup)',\n",
       "       'enzyme inhibitors: matrix metalloprotease inhibitors',\n",
       "       'aldosterone antagonists (spironolactone type)',\n",
       "       'antithelmintics (tibendazole type)',\n",
       "       'quaternary ammonium derivatives; atropine derivatives',\n",
       "       'antimetabolites (folic acid derivatives)',\n",
       "       'cholecystokinin (CCK) antagonists, antiulcer, anxiolytic agent',\n",
       "       'iodine-containing contrast media; enzyme inhibitors: folate hydrolose (also called Prostate Specific Membrane Antigen (PSMA) or glutamate carboxypeptidase II) inhibitors',\n",
       "       'peptides: peptides with muranic acid present',\n",
       "       'muscarinic receptor agonists/antagonists',\n",
       "       'antiviral (arildone derivatives)',\n",
       "       'anti-inflammatory agents (anilinonicotinic acid derivatives)',\n",
       "       'anthelmintics (undefined group)', 'arotinoid derivatives',\n",
       "       'Monoamine oxidase (MAO) inhibitors, type B',\n",
       "       'antineoplastics, anthraquinone derivatives',\n",
       "       'hepatoprotectives (timonacic type); iodine-containing contrast media',\n",
       "       \"cholinergic agonists (arecoline derivatives used in treatment of Alzheimer's disease\",\n",
       "       'Fc fusion protein; steriodal compounds acting on progesterone receptors (excluding-gest- compounds); antibacterials, pristinamycin derivatives',\n",
       "       'anti-inflammatory analgesics (phenylbutazone type) Note: INN Stem is-buzone',\n",
       "       'enzyme inhibitors: proteolytic enzyme inhibitors',\n",
       "       'antiulcer agents (benzimidazole derivatives): proton pump inhibitors',\n",
       "       'peptides: parathyroid hormone related peptide',\n",
       "       'enzyme inhibitors: tryptophan hydroxylase inhibitors',\n",
       "       'antivirals (quinoline derivatives)',\n",
       "       'hypnotics/sedatives (zolpidem type)',\n",
       "       'narcotic analgesics (fentanyl derivatives)',\n",
       "       'hepatoprotectives (timonacic type)',\n",
       "       'antibacterials (trimethoprim type), dihydrofolate reductase (DHFR) inhibitors',\n",
       "       'steriodal compounds acting on progesterone receptors (excluding-gest- compounds); selective progesterone receptor modulators (SPRM)',\n",
       "       'tranquilizers/antiepileptics (propanediol and pentanediol groups)',\n",
       "       'antineoplastics; antivirals (zidovudine group) (exception: edoxudine)',\n",
       "       'nonsteroidal ligand for the progesterone receptor',\n",
       "       'anticoagulants (dicumarol type)',\n",
       "       'steroids (androgens, anabolics)',\n",
       "       'tachykinin (neurokinin) receptor antagonists: NK1 receptor antagonists',\n",
       "       'antibacterial antibiotics, carbapenem derivatives',\n",
       "       'renin inhibitors; narcotic agonists/antagonists (normorphine type)',\n",
       "       'endopeptidase inhibitors: ACE (angiotensin-1 converting enzyme) and NEP (neutral endopeptidase) inhibitors',\n",
       "       'antimalarials (artemisin derivatives)', 'estrogens; progestins',\n",
       "       'class III antiarrhythmic agents',\n",
       "       'systemic antifungals (miconazole type); phosphoro-derivatives',\n",
       "       'antiparasitics (salicylanilide derivatives)',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines): leukotriene receptor antagonists',\n",
       "       'antihyperlipidaemics, acyl CoA: cholesterol acyltransferase (ACAT) inhibitors',\n",
       "       'progestins; topical steroids (acetal derivatives)',\n",
       "       'serotonin 5-HT2 receptor antagonists; atropine derivatives',\n",
       "       'phosphoro-derivatives; antivirals; antineoplastics (uridine derivatives)',\n",
       "       'antisense oligonucleotides',\n",
       "       'Fc fusion protein; alpha-adrenoceptor antagonists (benzodioxane derivatives)',\n",
       "       'serotonin 5-HT2 receptor antagonists',\n",
       "       'prednisone and prednisolone derivatives',\n",
       "       'enzyme inhibitors: nitric oxide synthase inhibitors',\n",
       "       'antithrombotics (thromboxane A2 receptor antagonists)',\n",
       "       'antibiotics obtained from Streptomyces kanamyceticus (related to kanamycin)',\n",
       "       'leukotriene receptor antagonists (treatment of inflammatory skin disorders)',\n",
       "       'antivirals: carbocyclic nucleosides',\n",
       "       'phosphodiesterase type 5 inhibitors (PDE5) inhibitors: (PDE5) inhibitors containing a sulfonamide moiety',\n",
       "       'platelet-activating factor antagonists',\n",
       "       'tyrosine kinase inhibitors: tyrosine kinase inhibitors; janus kinase inhibitors',\n",
       "       'vasodilators (undefined group); phosphoro-derivatives',\n",
       "       'prehormones or hormone-release stimulating peptides: thyrotropin releasing hormone analogues',\n",
       "       'antineoplastics (aromatase inhibitors)',\n",
       "       'quaternary ammonium derivatives: neuromuscular blocking agents',\n",
       "       'antiparasitics (ivermectin type)',\n",
       "       'fluoroquinolone derivatives, nonantibacterial indication (e.g., antineoplastic antibiotics)',\n",
       "       'diuretics (furosemide type)', 'ferrorportin receptor modulators',\n",
       "       'antacid aluminum salts',\n",
       "       'antiallergic respiratory tract drugs (xanoxic acid derivatives)',\n",
       "       'anticonvulsants; antimigraine (benzoylamino-benzpyran derivatives)',\n",
       "       'non-steroid antiandrogens',\n",
       "       'non-prostanoid prostaglandin receptor agonists',\n",
       "       'antihypertensives (prazosin type)',\n",
       "       'anti-inflammatory agents (anthranilic acid derivatives) and their salts or esters',\n",
       "       'phosphoro-derivatives; antihypertensives (ACE inhibitors): antihypertensives (ACE inhibitors), diacid analogs',\n",
       "       'muscarinic receptor antagonists',\n",
       "       'antibacterials (quinoline dioxide derivatives)',\n",
       "       'antidiabetics; class III antiarrhythmic agents',\n",
       "       'antimycobacterials (diaminodiphenylsulfone derivatives)',\n",
       "       'diuretics (thiazide derivatives)', 'beta-lactamase inhibitors',\n",
       "       'cholesterol sequestrants (glycosides)',\n",
       "       'phosphoro-derivatives; cytostatic phospholipid derivatives',\n",
       "       'iodine-containing contrast media; sulpride derivatives and analogs',\n",
       "       'enzyme inhibitors: aldose-reductase inhibitors; enzyme inhibitors: antihyperlipidemics (HMG-CoA inhibitors)',\n",
       "       'antimycotics (pyridone derivatives)', 'androgens',\n",
       "       'anti-inflammatory agents (isoxicam type)',\n",
       "       'progestins; steroids (androgens, anabolics); antiandrogens',\n",
       "       'antivirals: HIV protease inhibitors (saquinavir type)',\n",
       "       'caspase inhibitors',\n",
       "       'uracil type antineoplastics; uracil derivatives used as thyroid antagonists and as antineoplastics',\n",
       "       'monobactam antibiotics', '5-lipoxygenase inhibitors',\n",
       "       'selective serotonin reuptake inhibitors (SSRI)',\n",
       "       'hypoglycemics (phenformin type)',\n",
       "       'antianxiety agents (diazepam type); phosphoro-derivatives',\n",
       "       'aromatase inhibitors (imidazole/triazole derivatives); antimicrobials (sulfonamides derivatives)',\n",
       "       'ionotropic non-NMDA (N-methyl D-aspartate) glutamate receptors [AMPA (alpha-amino-3-hydroxy-5-methyl-4-isoxazolepropionic acid receptor) and/or KA (kainite antagonist) receptors]; endopeptidase inhibitors: ACE (angiotensin-1 converting enzyme) and NEP (neutral endopeptidase) inhibitors',\n",
       "       'general inhalation anesthetics (halogenated alkane derivatives)',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines): leukotriene biosynthesis inhibitors',\n",
       "       'phosphoro-derivatives; antivirals: non-nucleoside reverse transcriptase inhibitors, not benzoxazinone derivatives',\n",
       "       'peptides: peptides used as pulmonary surfactants',\n",
       "       'antihyperlipidemics (clofibrate type); anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'vasoconstrictors (vasopressin derivatives)',\n",
       "       'rapidly accelerated fibrosarcoma (RAF) kinase inhibitors',\n",
       "       'narcotic agonists/antagonists (normorphine type); enzyme inhibitors: aldose-reductase inhibitors',\n",
       "       'vitamin D analogs',\n",
       "       'prehormones or hormone-release stimulating peptides: growth hormone-release stimulating peptides',\n",
       "       'calcium metabolism regulators',\n",
       "       'diuretics (ethacrynic acid derivatives)',\n",
       "       'synthetic analogs of the dolastatin series',\n",
       "       'antiarrhythmics (disopyramide derivatives)',\n",
       "       'lipoprotein lipase activators',\n",
       "       'inhibitor of the menin-MLL (KMT2A) receptor',\n",
       "       'reactive nitrogen and/or oxygen species redox modulators',\n",
       "       'prehormones or hormone-release stimulating peptides',\n",
       "       'tyrosine kinase inhibitors',\n",
       "       'phlorozin derivatives, phenolic glycosides',\n",
       "       'enzyme inhibitors: xanthine oxidase/dehydrogenase inhibitors',\n",
       "       'antidepressants (fluoxetine type)',\n",
       "       'microsomal triglyceride transfer protein (MTP) inhibitors',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines): leukotriene receptor antagonists; narcotic agonists/antagonists (normorphine type)',\n",
       "       'cardiotonics (amrinone type)',\n",
       "       'non-steriodal anti-inflammatory agents, fenbufen derivatives',\n",
       "       'antirheumatics (lobenzarit type)',\n",
       "       'benzodiazepine receptor agonists, partial agonists, inverse agonists (quinolone derivatives)',\n",
       "       'estrogens; estrogen antagonists',\n",
       "       'quaternary ammonium derivatives; quaternary ammonium derivatives; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'renin inhibitors', 'antiseptics (dapabutan type)',\n",
       "       'phosphoro-derivatives; isophosphoramide mustard derivatives',\n",
       "       'antivirals: serine protease inhibitors',\n",
       "       'peroxisome proliferator activiating receptor (PPAR) agonists (thiazolidene derivatives)',\n",
       "       'phosphoro-derivatives; vinca alkaloids',\n",
       "       'analgesics (mixed opiate receptor agonists/antagonists); anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'catechol-O-methyltransferase (COMT) inhibitors',\n",
       "       'cortisone derivatives',\n",
       "       'antivirals/antiparkinsonians (adamantane derivatives)',\n",
       "       'protein tyrosine phosphatase 1B inhibitors',\n",
       "       'thalidomide derivatives',\n",
       "       'antianxiety, antidepressant inhibitor of norepinephrine and dopamine re-uptake',\n",
       "       'ileal bile acid transporter (IBAT) inhibitors',\n",
       "       'dopamine receptor agonists',\n",
       "       'tachykinin (neurokinin) receptor antagonists: NK3 receptor antagonists',\n",
       "       'antivirals; antivirals, chemokine receptor (CCR) antagonists',\n",
       "       'antimigraine, serotonin 5-HT1 receptor agonists',\n",
       "       'antiestrogens of the clomifene and tamoxifen groups; atropine derivatives',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines); Fc fusion protein',\n",
       "       'phosphoro-derivatives; phosphoro-derivatives; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'antibiotics (Streptomyces strain); antivirals',\n",
       "       'multidrug resistance inhibitors: acridine carboxamide derivatives',\n",
       "       'antidiabetics; peptides',\n",
       "       'estrogens; antineoplastics (chloroethylamine derivatives)',\n",
       "       'antiviral (arildone derivatives); phosphoro-derivatives',\n",
       "       'anthelmintics (undefined group); anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'immunomodulators: napthyridine analogs',\n",
       "       'steroids (not prednisolone derivatives); steroids (androgens, anabolics)',\n",
       "       'hormone-release inhibiting peptides', 'oxytocin derivatives',\n",
       "       'erythromycin derivatives lacking antibiotic activity; motilin agonists',\n",
       "       'serotonin receptor antagonists and partial agonists',\n",
       "       'hepatoprotectants, di-isopropyl (1,3-dithiol-2-ylidene) propanedioate and related compounds',\n",
       "       'vitamin D analogs; ergot alkaloid deriviatives',\n",
       "       'peptides: neuropeptide Y (NPY) receptors and analogues',\n",
       "       'tyrosine kinase inhibitors: tropomyosin receptor kinase (TRK) inhibitors',\n",
       "       'cephalosporins; quaternary ammonium derivatives; quaternary ammonium derivatives',\n",
       "       'steriodal compounds acting on progesterone receptors (excluding-gest- compounds); steroids (androgens, anabolics); testosterone reductase inhibitors',\n",
       "       'phosphatidylinositol 3-kinase (PI3K) inhibitors',\n",
       "       'amfetamine derivatives; iodine-containing contrast media',\n",
       "       'blood substitutes and/or diagnostics (perfluorochemicals)',\n",
       "       'ryanodine receptor modulators',\n",
       "       'quaternary ammonium derivatives; quaternary ammonium derivatives; atropine derivatives',\n",
       "       'phosphoro-derivatives; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'guanaline cyclase activators',\n",
       "       'anabolic steroids; platelet aggregation inhibitors, primarily platelet P2Y12 receptor antagonists',\n",
       "       'cortisone derivatives; topical steroids (acetal derivatives)',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines): integrin antagonists',\n",
       "       'cephalosporins; endopeptidase inhibitors',\n",
       "       'Fc fusion protein; antihyperlipidaemics, acyl CoA: cholesterol acyltransferase (ACAT) inhibitors',\n",
       "       'sulpride derivatives and analogs; atropine derivatives',\n",
       "       'tricyclic compounds; Rauwolfia alkaloid derivatives',\n",
       "       'phosphoro-derivatives; dopaminergics (butopamine type)',\n",
       "       'nonsteroidal anti-inflammatory agents (tenidap type)',\n",
       "       'maytansinoid derivatives', 'cathespin inhibitors',\n",
       "       'aptamers, classical and mirror; peptides',\n",
       "       'cephalosporins; quaternary ammonium derivatives',\n",
       "       'tranquilizers/antiepileptics (propanediol and pentanediol groups); barbituric acid derivatives',\n",
       "       '5-lipoxygenase-activating protein (FLAP) inhibitors',\n",
       "       'proton pump inhibitors, not dependent on acid activation',\n",
       "       'hypoglycemics (phenformin type); phosphoro-derivatives',\n",
       "       'barbituric acid derivatives; vinca alkaloids',\n",
       "       'aromatase inhibitors (imidazole/triazole derivatives)',\n",
       "       'phenylpyridine vasodilators; Fc fusion protein; tricyclic compounds',\n",
       "       'analgesics (mixed opiate receptor agonists/antagonists); steroids (not prednisolone derivatives)',\n",
       "       'endopeptidase inhibitors; vinca alkaloids',\n",
       "       'complement receptor antagonists/complement inhibitors',\n",
       "       'estrogens; endopeptidase inhibitors',\n",
       "       'enzyme inhibitors: glutaminase inhibitor',\n",
       "       'cyclosporin derivatives',\n",
       "       'antivirals/antiparkinsonians (adamantane derivatives); growth hormone derivatives',\n",
       "       'steriodal compounds acting on progesterone receptors (excluding-gest- compounds); angiotensin II receptor antagonists',\n",
       "       'iodine-containing contrast media; steroids (androgens, anabolics); bronchodilators (phenethylamine derivatives)',\n",
       "       'combined alpha and beta blockers; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'enzymes',\n",
       "       'class III antiarrhythmic agents; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'aptamers, classical and mirror', 'oxazolidinone antibacterials',\n",
       "       'immunostimulants',\n",
       "       \"acetylcholinesterase inhibitors used in the treatment of Alzheimer's disease\",\n",
       "       'gonadotropin releasing hormone (GnRH) receptor antagonists (nonpeptide)',\n",
       "       'histamine H3 receptor antagonists',\n",
       "       'cortisone derivatives; steroids (androgens, anabolics); antiandrogens',\n",
       "       'tyrosine kinase inhibitors: tyrosine kinase inhibitors; MEK (MAPK kinase) inhibitors',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives); vinca alkaloids',\n",
       "       'cannabinol derivatives: CB (e.g., CB, CB2, etc.) cannabinoid receptor antagonists',\n",
       "       'thrombopoetin agonists',\n",
       "       'tricyclic histaminic-H1 receptor antagonists, loratadine derivatives (formerly-tadine)',\n",
       "       'enzyme inhibitors: acetyl CoA carboxylase (ACC) inhibitors',\n",
       "       'CRF-1 (corticotropin releasing factor) receptor antagonists',\n",
       "       'phosphoro-derivatives; antivirals: RNA polymerase (NS5B) inhibitors',\n",
       "       'narcotic agonists/antagonists (normorphine type); tricyclic compounds',\n",
       "       'antineoplastics, taxane derivatives',\n",
       "       'enkephalin agonists (various indications)',\n",
       "       'peptides: inhibition of growth hormone release; chelating agents',\n",
       "       'dopamine receptor agonists; phosphoro-derivatives',\n",
       "       'inhibitors of the ADAM (A disintegrin and A metalloprotease) enzymes (sheddases)',\n",
       "       'antineoplastics (arabinofuranosyl derivatives)',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines): thromboxane A2 receptor antagonists',\n",
       "       'anthelmintics (undefined group): 2-deoxoparaherquamide A derivatives',\n",
       "       'heat-shock protein inducers, bimoclomal type',\n",
       "       'Fc fusion protein; phospholipase A2 inhibitors',\n",
       "       'immunomodulators; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'angiogenesis inhibitors',\n",
       "       'steriodal compounds acting on progesterone receptors (excluding-gest- compounds); antibacterials, pristinamycin derivatives',\n",
       "       'enzyme inhibitors: histone deacetylase inhibitors',\n",
       "       'Amyloid beta (Abeta) binders',\n",
       "       'enzyme inhibitors: beta secretase inhibitors',\n",
       "       'nociceptin opion receptor (NOP) agonists',\n",
       "       'phosphodiesterase 10 (PDE10) inhibitors',\n",
       "       'antihrombotic indirect selective synthetic factor Xa inhibitors',\n",
       "       'uracil type antineoplastics',\n",
       "       'antidiabetic, sodium glucose co-transporter 2 (SGLT2) inhibitors, not phlorozin derivatives',\n",
       "       'selective melatonin receptor agonist',\n",
       "       'adenosine A receptor agonists', 'oxytocin antagonists',\n",
       "       'antibiotics, DNA gyrase and topoisomerase IV inhibitors',\n",
       "       'proteozome inhibitors', 'monoamine transport inhibitors',\n",
       "       'antivirals: respiratory syncytial virus (RSV) fusion inhibitors',\n",
       "       'pteridine derivatives',\n",
       "       'nicotinic acetylcholine receptor partial agonists/agonists',\n",
       "       'enzyme inhibitors: anti-fibrotic/cyclophilin inhibitor',\n",
       "       'diuretics (piretanide type)',\n",
       "       'phosphoro-derivatives; antibiotics (Streptomyces strain)',\n",
       "       'chemokine (C-X-C motif) receptor 2 (CXCR2) modulators',\n",
       "       'bcl-2 (B-cell lymphoma 2) inhibitors', 'kinesin inhibitors',\n",
       "       'deuterated compounds',\n",
       "       'endopeptidase inhibitors: neutral endopeptidase inhibitors possessing additional endothelin converting enzyme inhibitory activity',\n",
       "       'antivirals: inhibitor of COVID 3CL protease',\n",
       "       'antineoplastic thymidylate synthetase inhibitors',\n",
       "       'antivirals: glucosidase inhibitors',\n",
       "       'mutated isocitrate dehydrogenase (IDH) inhibitors',\n",
       "       'filamin A binders',\n",
       "       'cyclin dependent kinase inhibitors (formerly-cidib)',\n",
       "       'tyrosine kinase inhibitors: epidermal growth factor receptor (EGFR) inhibitors',\n",
       "       'spleen tyrosine kinase (Syk) inhibitors',\n",
       "       'antidepressants (fluoxetine type): thioether derivatives',\n",
       "       'metabotropic glutamate receptor antagonists, negative allosteric modulators',\n",
       "       'trace amine associated receptor (TAAR) agonists',\n",
       "       'antivirals: inhibitor of functional assembly of viral core proteins',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives); anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'antibiotics (oxacefalosporanic acid derivatives)',\n",
       "       'camphorsulfonic acid derivatives used as UVA sunscreens',\n",
       "       'enzyme inhibitors: hypoxia-inducible factor (HIF) prolyl hydroxylase inhibitors',\n",
       "       'aptamers, classical and mirror; peptides: gap junction protein channel modulators',\n",
       "       'extra-terminal motif (BET) family of bromodomain (BRD) proteins',\n",
       "       'immunomodulators: mitogen-activated protein (MAP) kinase inhibitors',\n",
       "       'benzodiazepine receptor agonists/antagonists; iodine-containing contrast media',\n",
       "       'aptamers, classical and mirror; vasopressin receptor antagonists',\n",
       "       'enzyme inhibitors: gamma secretase inhibitors',\n",
       "       'peptides: platelet aggregation inhibitor (GPIIb/IIIa receptor antagonist)',\n",
       "       \"tyrosine kinase inhibitors: tyrosine kinase inhibitors Bruton's (Btk) inhibitors\",\n",
       "       'antivirals: non-nucleoside reverse transcriptase inhibitors, not benzoxazinone derivatives',\n",
       "       'dipeptidyl aminopeptidase-IV inhibitors',\n",
       "       'enzyme inhibitors: histone deacetylase inhibitors; vinca alkaloids',\n",
       "       'immunomodulators: immunomodulators, toll-like receptor agonists',\n",
       "       'cholesterol ester transfer protein inhibitors',\n",
       "       'skeletal muscle troponin activator',\n",
       "       'narcotic agonists/antagonists (normorphine type); antihypertensives (ACE inhibitors)',\n",
       "       'inhibitor of apoptosis protein (IAP)',\n",
       "       'tachykinin (neurokinin) receptor antagonists: NK2 receptor antagonists',\n",
       "       'heat shock protein inhibitors',\n",
       "       'estrogen receptor beta agonist; cannabinol derivatives',\n",
       "       'ionotropic non-NMDA (N-methyl D-aspartate) glutamate receptors [AMPA (alpha-amino-3-hydroxy-5-methyl-4-isoxazolepropionic acid receptor) and/or KA (kainite antagonist) receptors]; poly-ADP-ribose polymerase inhibitors',\n",
       "       'glucocorticoid receptor agonists ( not glucocorticoids); phosphoro-derivatives',\n",
       "       'antivirals: RNA polymerase (NS5B) inhibitors',\n",
       "       'peptides: melanocortin receptor agonists',\n",
       "       'melanocortin-1 receptor (MC1R) agonists',\n",
       "       'enzyme inhibitors: glucosyltransferase inhibitors',\n",
       "       'adenosine receptor antagonists',\n",
       "       'antivirals: nonstructural protein 5A (NS5A) inhibitors',\n",
       "       'peptides: atrial natriuretic type substances',\n",
       "       'hedgehog signaling inhibitors',\n",
       "       'antivirals: viral capsid inhibitors',\n",
       "       'Fc fusion protein; antibiotics (Streptomyces strain)',\n",
       "       'transient receptor potential antagonists; vanilloid 1 receptor antagonists',\n",
       "       'poly-ADP-ribose polymerase inhibitors',\n",
       "       'dopamine D1-like receptors (e.g., D1 and D5 subtypes) agonists/partial agonists and potentiators',\n",
       "       'local anesthetics; penicillins',\n",
       "       'potassium channel antagonists: mixed sodium/potassium channel blockers',\n",
       "       'iodine-containing contrast media; atropine derivatives',\n",
       "       'Kirsten rat sarcoma (KRAS) inhibitors',\n",
       "       'antiarrhythmics: indicates high iodine content',\n",
       "       'phosphodiesterase type 9 (PDE9) inhibitors',\n",
       "       'calcitonin gene-related peptide receptor antagonists',\n",
       "       'iodine-containing contrast media; diuretics (sulfamoylbenzoic acid derivatives)',\n",
       "       'deuterated compounds; tyrosine kinase inhibitors',\n",
       "       'antimicrobial, bactericidal permeability increasing polypeptide',\n",
       "       'phosphoro-derivatives; antiepileptics (hydantoin derivatives)',\n",
       "       'extracellular signal regulated kinase (ERK) types 1 and 2 inhibitors (formerly-xertinib)',\n",
       "       'phosphoro-derivatives; tyrosine kinase inhibitors',\n",
       "       'antivirals: adenosine analogs acting as RNA polymerase inhibitors',\n",
       "       'enzyme inhibitors: gastrointestinal lipase inhibitors',\n",
       "       'vitamin D analogs; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'antagonist of the glucagon receptor',\n",
       "       'antivirals: neuraminidase inhibitors',\n",
       "       'aptamers, classical and mirror; antihypertensives (ACE inhibitors)',\n",
       "       'antineoplastics (rebeccamycin derivatives)',\n",
       "       'antivirals: integrase inhibitors',\n",
       "       'antivirals: treatments of SARS-CoV-2 infection, inhibiting viral RNA replication',\n",
       "       'aptamers, classical and mirror; antibiotics (Streptomyces strain)',\n",
       "       'peptides: (prostate-specific membrane antigen) PSMA binding peptide; chelating agents',\n",
       "       'survival motor neurin SMN1 and SMN2 splicing modulators',\n",
       "       'cystic fibrosis transmembrane regulator (CFTR) protein modulators',\n",
       "       'vasodilators (undefined group); antiarrhythmic/antianginal/antihypertensive agents, phthalazine like structure',\n",
       "       'glucocorticoid receptor antagonists (not glucocorticoids)',\n",
       "       'opener of large conductance calcium-activated (maxi-K) potassium channels',\n",
       "       'vancomycin-related compounds',\n",
       "       'retinol derivatives; tyrosine kinase inhibitors',\n",
       "       'calcium receptor antagonists',\n",
       "       'aptamers, classical and mirror; pteridine derivatives',\n",
       "       'antineoplastics, apoptosis inducing sulfonamide',\n",
       "       'multidrug resistance inhibitors: ciclosporin D derivatives',\n",
       "       'serum amyloid P (SAP) inhibitors', 'polymers',\n",
       "       'peptides: glucagon-like peptide (GLP) analogs',\n",
       "       'spermine and/or homospermine derivatives or analogs',\n",
       "       'boroles, boron compounds',\n",
       "       'androgens; steroids (androgens, anabolics); antiandrogens',\n",
       "       'allosteric modulators of small-conductance calcium activated potassium (SK) channels; viral capsids',\n",
       "       'maytansinoid derivatives; peptides: inhibition of growth hormone release',\n",
       "       'antibiotics (carbacephem derivatives)',\n",
       "       'N-methyl D-aspartate (NMDA) receptor antagonists (N2B subunit)',\n",
       "       'sodium-hydrogen antiporter (also called sodium-proton exchanger) inhibitors',\n",
       "       'deuterated compounds; tyrosine kinase inhibitors: tyrosine kinase inhibitors; janus kinase inhibitors',\n",
       "       'Chemokine receptor (CCR) antagonists, not antivirals',\n",
       "       'cardiac myosin activator',\n",
       "       'tricyclic compounds; antibiotics (rifamycin derivatives)',\n",
       "       'vasodilators (undefined group); prostaglandins',\n",
       "       'glucagon-like peptide receptor agonists',\n",
       "       'enzyme inhibitors: glutathione-s- transferase inhibitors',\n",
       "       'androgens; steroids (not prednisolone derivatives)',\n",
       "       'antihistaminics (histamine-H1 receptor antagonists); vinca alkaloids',\n",
       "       'enzyme inhibitors: small ubiquitin-like modifier inhibitor',\n",
       "       'multidrug resistance inhibitors: quinoline derivatives',\n",
       "       'vitamin D analogs; antibiotics (Streptomyces strain)',\n",
       "       'enzyme inhibitors: autotaxin inhibitors',\n",
       "       'prostaglandin receptor antagonists, non prostinoid structure',\n",
       "       'PPAR delta receptor agonists', 'pyruvate kinase activators',\n",
       "       'analgesics (mixed opiate receptor agonists/antagonists); beta-blockers (propranolol type)',\n",
       "       'kappa opiod receptor antagonists', 'CXCR receptor antagonists',\n",
       "       'enzyme inhibitors: 1) kallikrein inhibitors or 2) serine protease inhibitors',\n",
       "       'deuterated compounds; antiasthmatics (bradykinin antagonists)',\n",
       "       'nitrogen oxide (eg., nitrxyl and nitric oxide) donors; prostaglandins',\n",
       "       'toll-like receptor antagonists',\n",
       "       'ionotropic non-NMDA (N-methyl D-aspartate) glutamate receptors [AMPA (alpha-amino-3-hydroxy-5-methyl-4-isoxazolepropionic acid receptor) and/or KA (kainite antagonist) receptors]: modulators',\n",
       "       'multidrug resistance inhibitors: pipecolic acid derivatives',\n",
       "       'anticonvulsants; antimigraine (benzoylamino-benzpyran derivatives); cannabinol derivatives',\n",
       "       'selective androgen receptor modulator (non-steroidal)',\n",
       "       'mitochondrial benzodiazepine receptor (MBR) selective, partial, or inverse agonists (purine derivatives)',\n",
       "       'aldehyde traps',\n",
       "       'antineoplastics (mitotic inhibitors; tubulin binders); cannabinol derivatives',\n",
       "       'histamine H4 receptor antagonists',\n",
       "       'antiarrhythmics (ajmaline derivatives)',\n",
       "       'enzyme inhibitors: indoleamine 2, 3-dioxygenase 1 (IDO1) inhibitors',\n",
       "       'antagonists of the transient receptor potential channel (TRP), melatsatin (TRPM) type; transient receptor potential antagonists',\n",
       "       'peptides: peptides with prostate cancer indications',\n",
       "       'urotensin receptor antagonists',\n",
       "       'phosphoro-derivatives; tachykinin (neurokinin) receptor antagonists: NK1 receptor antagonists',\n",
       "       'antiarrhythmics (ajmaline derivatives); quaternary ammonium derivatives',\n",
       "       'phosphodiesterase type 9 (PDE9) inhibitors; vinca alkaloids',\n",
       "       'antineoplastics, pro-apoptotic cGMP phosphodiesterase inhibitors, sulindac metabolites',\n",
       "       'steriodal compounds acting on progesterone receptors (excluding-gest- compounds)',\n",
       "       'benzoporphyrin derivatives', 'depsipeptide derivatives',\n",
       "       'retinoid-related orphan receptor-Î³t (RORÎ³t) antagonists/inverse agonists',\n",
       "       'mycolic acid inhibitors',\n",
       "       'peptides: gastrointestinal functions; normalizing agent',\n",
       "       'antimyloidotics', 'peptides: antiviral peptides; antivirals',\n",
       "       'dopamine D2D3 receptor modulators',\n",
       "       'anticoagulants (hirudin type)',\n",
       "       'enzyme inhibitors: partial fatty acid oxidation (pFOX) inhibitor',\n",
       "       'serotonin 5-HT3 receptor antagonists; atropine derivatives',\n",
       "       'tricyclic compounds; atropine derivatives',\n",
       "       'phosphoro-derivatives; oxazolidinone antibacterials',\n",
       "       'nitrogen oxide (eg., nitrxyl and nitric oxide) donors',\n",
       "       'enzyme inhibitors: dopamine b-hydrolase (DBH) inhibitors',\n",
       "       'enzyme inhibitors: Cluster of differentiation (CD) inhibitors (alternative: CD73 inhibitors)',\n",
       "       'tachykinin (neurokinin) receptor antagonists: neurotensin receptor antagonists',\n",
       "       'receptor for advanced glycation endproducts (RAGE) inhibitors',\n",
       "       'ergot alkaloid deriviatives; phlorozin derivatives, phenolic glycosides',\n",
       "       'phosphoro-derivatives; immunomodulators',\n",
       "       'ubiquitin ligase inhibitors',\n",
       "       'silent mating type information regulator (SIRT1) receptor agonists',\n",
       "       'antihypertensives (guanidine deriviatives); alpha-adrenoceptor antagonists (benzodioxane derivatives)',\n",
       "       'antineoplastic, acylfulvene derivatives',\n",
       "       'cephalosporins; vancomycin-related compounds',\n",
       "       'farnesyl transferase inhibitor', 'epothilones',\n",
       "       'estrogen receptor beta agonist',\n",
       "       'cephalosporins; opioid receptor antagonists',\n",
       "       'serotonin 5-HT receptor agonists, not principally 5-HT2',\n",
       "       'E-selectin antagonists', 'anabolic steroids; enzyme inhibitors',\n",
       "       'vasodilators (undefined group): Rho protein kinase (ROCK) inhibitors',\n",
       "       'organoarsenic compounds',\n",
       "       'glucocorticoid receptor agonists ( not glucocorticoids)',\n",
       "       'alpha1- and alpha2- adrenoreceptor agonists',\n",
       "       'Fc fusion protein; antivirals: non-nucleoside reverse transcriptase inhibitors; benzoxazinone derivatives',\n",
       "       'growth hormone derivatives', 'antibiotics (polyene derivatives)',\n",
       "       'heat shock protein inhibitors; narcotic agonists/antagonists (normorphine type)',\n",
       "       'cannabinol derivatives; enzyme inhibitors: beta secretase inhibitors',\n",
       "       'antagonists of heparin and/or low molecular weight heparins',\n",
       "       'antivirals: cyclosporine derivatives',\n",
       "       'narcotic agonists/antagonists (normorphine type); antihypertensives (ACE inhibitors): antihypertensives (ACE inhibitors), diacid analogs',\n",
       "       'neurotoxic protein translation inhibitors',\n",
       "       'antipsychotic with binding activity on sertonin (5-HT2A) and dopamine (D2) receptors',\n",
       "       'warfarin analogs',\n",
       "       'vinca alkaloids; WNT (wingless-type mouse mammary tumor virus) pathway inhibitors',\n",
       "       'enzyme inhibitors: squalene synthase inhibitors',\n",
       "       'antiarrhythmic/antianginal/antihypertensive agents, phthalazine like structure; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'antineoplastics (chloroethylamine derivatives); prednisone and prednisolone derivatives',\n",
       "       'coagulation factor XIa inhibitors',\n",
       "       'allosteric cardiac myosin modulators', 'FimH (E.coli) blockers',\n",
       "       'ecteinascidin derivatives',\n",
       "       'enzyme inhibitors: folate hydrolose (also called Prostate Specific Membrane Antigen (PSMA) or glutamate carboxypeptidase II) inhibitors',\n",
       "       'antiviral (arildone derivatives); cephalosporins',\n",
       "       'PEGylated compounds, covalent attachment of macrogol (polyethylene glycol) polymer; tyrosine kinase inhibitors',\n",
       "       'prehormones or hormone-release stimulating peptides; antineoplastic antibiotics (daunorubicin type)',\n",
       "       'aptamers, classical and mirror; ergot alkaloid deriviatives',\n",
       "       'liver X nuclear hormone (LXR) receptor agonists',\n",
       "       'systemic antifungals (miconazole type); Fc fusion protein',\n",
       "       'phosphoro-derivatives; antineoplastics; antivirals (zidovudine group) (exception: edoxudine)',\n",
       "       'cytochrome bc1 complex inhibitors in mycobacterium tuberculosis',\n",
       "       'enzyme inhibitors: D-amino acid oxidase (DAAO) inhibitors',\n",
       "       'Kirsten rat sarcoma (KRAS) inhibitors; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'phosphoro-derivatives; antivirals',\n",
       "       'enzyme inhibitors: DNA polymerase',\n",
       "       'phosphatidylinositol 3-kinase (PI3K) inhibitors; narcotic agonists/antagonists (normorphine type)',\n",
       "       \"Src homology 2-containing inositol-5' phosphatase 1 (SHIP1) activators\",\n",
       "       'angiogenesis inhibitors; small interfering RNA (siRNA)',\n",
       "       'histone acetyltransferase inhibitors, targeting the bromodomain acetyl transferase proteins adenoviral E1A-binding protein (p300) and cAMP-responsive element binding protein (CREB) binding protein (CBP)',\n",
       "       'inosine monophosphate dehydrogenase inhibitors',\n",
       "       'non-peptidic growth hormone secretagogues',\n",
       "       'epithelial sodium channel (ENaC) inhibitors, amiloride derivatives',\n",
       "       'enzyme inhibitors: fatty acid synthase inhibitors',\n",
       "       'phosphoro-derivatives; pteridine derivatives',\n",
       "       'calcium-sensing receptor (CaSR) agonists',\n",
       "       'PEGylated compounds, covalent attachment of macrogol (polyethylene glycol) polymer; antineoplastics (camptothecin derivatives)',\n",
       "       'glycine transporter/reuptake inhibitors',\n",
       "       'antivirals: antiviral, disrupts viral maturation',\n",
       "       'antibiotics (Streptomyces strain); antibiotics (rifamycin derivatives)',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives); cholinesterase inhibitors (physostigmine type)',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives); bronchodilators (phenethylamine derivatives)',\n",
       "       'vasodilators (undefined group): calcium channel blockers acting as vasodilators',\n",
       "       'kinase inhibitor, binds to myristoyl binding site',\n",
       "       'phosphodiesterase type 1 (PDE1) ihibitors',\n",
       "       'phosphoro-derivatives; antimycotics (pyridone derivatives)',\n",
       "       'ergot alkaloid deriviatives; vinca alkaloids',\n",
       "       'narcotic agonists/antagonists (normorphine type); 5-HT1A receptor agonists/antagonists acting primarily as neuroprotectors',\n",
       "       'antihypertensives (guanidine deriviatives); cannabinol derivatives',\n",
       "       '-',\n",
       "       'nucleoside antiviral or antineoplastic agents, cytarabine or azarabine derivatives; phosphoro-derivatives',\n",
       "       'anti-inflammatory agents (salicylic acid derivatives); peptides',\n",
       "       'thalidomide derivatives: zinc finger protein degradation inducing thalidomide derivatives',\n",
       "       'androgen receptor inhibitor, N-terminal domain inhibitor',\n",
       "       'antibiotics (Streptomyces strain); anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'antibiotics (tetracycline derivatives); phosphoro-derivatives; anti-inflammatory agents (salicylic acid derivatives)',\n",
       "       'dopamine D2 receptor modulator',\n",
       "       'phosphoro-derivatives; antivirals: HIV protease inhibitors (saquinavir type)',\n",
       "       'antiarrhythmic/antianginal/antihypertensive agents, phthalazine like structure; anti-inflammatory agents (salicylic acid derivatives); antimicrobials (sulfonamides derivatives)',\n",
       "       'thymidylate synthase inhibitors (benzindole derivatives)',\n",
       "       'retinol binding protein 4 (RBP4) antagonists',\n",
       "       'hypoxia-inducible factor (HIF) inhibitors',\n",
       "       'phosphoro-derivatives; cytostatic phospholipid derivatives; iodine-containing contrast media',\n",
       "       'inducers of ribosomal readthrough on nonsense mutation mRNA stop codons',\n",
       "       'neuropeptide Y5 antagonists',\n",
       "       'peptides: amylin derivatives or mimics',\n",
       "       'growth hormone derivatives; enzyme inhibitors: antihyperlipidemics (HMG-CoA inhibitors)',\n",
       "       'chelating agents',\n",
       "       'angiogenesis inhibitors; phosphoro-derivatives',\n",
       "       'antibiotics, diarylquinoline structure',\n",
       "       'Protease-activated receptor 1 (PAR1) antagonists',\n",
       "       'serotonin 5-HT1B receptor inhibitors',\n",
       "       'antifungals, triterpinoids',\n",
       "       'nootropic agents (learning/cognitive enhancers), purine derivatives',\n",
       "       'peptides: zonulin antagonists',\n",
       "       'enzyme inhibitors: heparin derivatives',\n",
       "       'deuterated compounds; non-steroid antiandrogens',\n",
       "       'neuronal apoptosis inhibitors; glyceraldehyde-3-phosphate dehydrogenase (GAPDH)',\n",
       "       'urate transporter inhibitors',\n",
       "       'analgesics (mixed opiate receptor agonists/antagonists); gabamimetics',\n",
       "       'peptides: oxyntomodulin analogs',\n",
       "       'selectin antagonists inhibiting all selectin types',\n",
       "       'enzyme inhibitors: lanosterol 14a-demethylase inhibitors',\n",
       "       'phosphoro-derivatives; antihypertensives (ACE inhibitors)',\n",
       "       'opioid receptor antagonists',\n",
       "       'farnesoid X receptor agonists; farnesoid X receptor agonists',\n",
       "       'tachykinin (neurokinin) receptor antagonists: NK3 receptor antagonists; vinca alkaloids',\n",
       "       'narcotic agonists/antagonists (normorphine type); bronchodilators (phenethylamine derivatives)',\n",
       "       'agonists of the Gardos channel, or the calcium activated potassium channel of intermediate conductance',\n",
       "       'thalidomide derivatives; narcotic agonists/antagonists (normorphine type)',\n",
       "       'phosphoro-derivatives; hepatocyte growth factor (HGF)-MET agonists',\n",
       "       'quaternary ammonium derivatives; vinca alkaloids',\n",
       "       'antiasthmatics/antiallergics (not acting primarily as antihistamines); vasodilators (undefined group)',\n",
       "       'antineoplastics (mitotic inhibitors; tubulin binders); phosphoro-derivatives',\n",
       "       'enzyme inhibitors: s-nitrosoglutathione reductase (GSNOR) inhibitors',\n",
       "       'deuterated compounds; narcotic antagonists/agonists (morphinan derivatives)',\n",
       "       'serotonin receptor agonists, primarily 5-HT2'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['USAN Definition'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the purpose of this exercise we will filter out some specific drugs based on their USAN definition. We picked these four mode of actions: *antivirals*, *analgesics*, *beta-blockers* and *anti-inflammatory* drugs. We will keep any data point that has these words in their description and then make four new super classes out of them stored in `application`. You are free to choose different classes and test the influence on the results below. Depending on the homogeneity of the groups the clustering will give very different results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_applications = ['antivirals', 'analgesics', 'beta-blockers','anti-inflammatory']\n",
    "# df matches one of the applications (not contains but exact match)\n",
    "df = df[df['USAN Definition'].str.contains('|'.join(drug_applications), case=False, na=False)]\n",
    "# add a column with the application combining all antivirals etc. into one class each\n",
    "df['application'] = df['USAN Definition'].apply(lambda x: [app for app in drug_applications if app in x.lower()][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Principal Component Analysis (PCA)\n",
    "Principal Component Analysis (PCA) is a technique used to emphasize variation and bring out strong patterns in a dataset. It's often used to make data easy to explore and visualize. It is also used for dimensionality reduction, which is useful when you have a lot of features describing your data. PCA is a linear transformation method that finds the directions (principal components) that maximize the variance in the data. These directions are orthogonal to each other and form a new coordinate system in which the data can be represented. The first principal component is the direction in which the data varies the most, the second principal component is the direction in which the data varies the second most, and so on. We will plot the first two principal components of the data and thus reduce the dimnesionality of the data (going from n to 2 dimensions). For more details have a look at [this blog post](https://towardsdatascience.com/principal-component-analysis-pca-explained-visually-with-zero-math-1cbf392b9e7d).\n",
    "\n",
    "We will perform PCA on Morgan fingerprints as features and make use of an the package `molplotly`, which allows plotting data interactively!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: molplotly in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (1.1.7)\n",
      "Requirement already satisfied: dash>=2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (2.10.0)\n",
      "Requirement already satisfied: werkzeug>=2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (2.2.3)\n",
      "Requirement already satisfied: jupyter-dash>=0.4.2 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (0.4.2)\n",
      "Requirement already satisfied: plotly>=5.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (5.22.0)\n",
      "Requirement already satisfied: rdkit-pypi>=2021.9.4 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (2022.9.5)\n",
      "Requirement already satisfied: pandas in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (2.1.4)\n",
      "Requirement already satisfied: ipykernel in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (6.29.2)\n",
      "Requirement already satisfied: nbformat in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from molplotly) (5.9.2)\n",
      "Requirement already satisfied: Flask<=2.2.3,>=1.0.4 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash>=2.0.0->molplotly) (2.2.3)\n",
      "Requirement already satisfied: dash-html-components==2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash>=2.0.0->molplotly) (2.0.0)\n",
      "Requirement already satisfied: dash-core-components==2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash>=2.0.0->molplotly) (2.0.0)\n",
      "Requirement already satisfied: dash-table==5.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash>=2.0.0->molplotly) (5.0.0)\n",
      "Requirement already satisfied: requests in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jupyter-dash>=0.4.2->molplotly) (2.31.0)\n",
      "Requirement already satisfied: retrying in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jupyter-dash>=0.4.2->molplotly) (1.3.4)\n",
      "Requirement already satisfied: ipython in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jupyter-dash>=0.4.2->molplotly) (8.22.1)\n",
      "Requirement already satisfied: ansi2html in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jupyter-dash>=0.4.2->molplotly) (1.9.1)\n",
      "Requirement already satisfied: nest-asyncio in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jupyter-dash>=0.4.2->molplotly) (1.6.0)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from plotly>=5.0.0->molplotly) (8.3.0)\n",
      "Requirement already satisfied: packaging in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from plotly>=5.0.0->molplotly) (23.2)\n",
      "Requirement already satisfied: numpy in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from rdkit-pypi>=2021.9.4->molplotly) (1.26.4)\n",
      "Requirement already satisfied: Pillow in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from rdkit-pypi>=2021.9.4->molplotly) (10.2.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from werkzeug>=2.0.0->molplotly) (2.1.5)\n",
      "Requirement already satisfied: appnope in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (0.1.4)\n",
      "Requirement already satisfied: comm>=0.1.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (0.2.1)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (1.8.1)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (8.6.0)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (5.7.1)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (0.1.6)\n",
      "Requirement already satisfied: psutil in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (5.9.8)\n",
      "Requirement already satisfied: pyzmq>=24 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (25.1.2)\n",
      "Requirement already satisfied: tornado>=6.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (6.4)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipykernel->molplotly) (5.14.1)\n",
      "Requirement already satisfied: fastjsonschema in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from nbformat->molplotly) (2.19.1)\n",
      "Requirement already satisfied: jsonschema>=2.6 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from nbformat->molplotly) (4.21.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from pandas->molplotly) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from pandas->molplotly) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from pandas->molplotly) (2024.1)\n",
      "Requirement already satisfied: Jinja2>=3.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Flask<=2.2.3,>=1.0.4->dash>=2.0.0->molplotly) (3.1.3)\n",
      "Requirement already satisfied: itsdangerous>=2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Flask<=2.2.3,>=1.0.4->dash>=2.0.0->molplotly) (2.2.0)\n",
      "Requirement already satisfied: click>=8.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Flask<=2.2.3,>=1.0.4->dash>=2.0.0->molplotly) (8.1.7)\n",
      "Requirement already satisfied: decorator in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (0.19.1)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (3.0.43)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (2.17.2)\n",
      "Requirement already satisfied: stack-data in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (0.6.3)\n",
      "Requirement already satisfied: exceptiongroup in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (1.2.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from ipython->jupyter-dash>=0.4.2->molplotly) (4.9.0)\n",
      "Requirement already satisfied: attrs>=22.2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jsonschema>=2.6->nbformat->molplotly) (23.2.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jsonschema>=2.6->nbformat->molplotly) (2023.12.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jsonschema>=2.6->nbformat->molplotly) (0.33.0)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jsonschema>=2.6->nbformat->molplotly) (0.18.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel->molplotly) (4.2.0)\n",
      "Requirement already satisfied: six>=1.5 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas->molplotly) (1.16.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from requests->jupyter-dash>=0.4.2->molplotly) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from requests->jupyter-dash>=0.4.2->molplotly) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from requests->jupyter-dash>=0.4.2->molplotly) (2.2.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from requests->jupyter-dash>=0.4.2->molplotly) (2024.2.2)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from jedi>=0.16->ipython->jupyter-dash>=0.4.2->molplotly) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from pexpect>4.3->ipython->jupyter-dash>=0.4.2->molplotly) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from prompt-toolkit<3.1.0,>=3.0.41->ipython->jupyter-dash>=0.4.2->molplotly) (0.2.13)\n",
      "Requirement already satisfied: executing>=1.2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from stack-data->ipython->jupyter-dash>=0.4.2->molplotly) (2.0.1)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from stack-data->ipython->jupyter-dash>=0.4.2->molplotly) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from stack-data->ipython->jupyter-dash>=0.4.2->molplotly) (0.2.2)\n",
      "Requirement already satisfied: dash==2.10 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (2.10.0)\n",
      "Requirement already satisfied: Flask<=2.2.3,>=1.0.4 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash==2.10) (2.2.3)\n",
      "Requirement already satisfied: Werkzeug<=2.2.3 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash==2.10) (2.2.3)\n",
      "Requirement already satisfied: plotly>=5.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash==2.10) (5.22.0)\n",
      "Requirement already satisfied: dash-html-components==2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash==2.10) (2.0.0)\n",
      "Requirement already satisfied: dash-core-components==2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash==2.10) (2.0.0)\n",
      "Requirement already satisfied: dash-table==5.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from dash==2.10) (5.0.0)\n",
      "Requirement already satisfied: Jinja2>=3.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Flask<=2.2.3,>=1.0.4->dash==2.10) (3.1.3)\n",
      "Requirement already satisfied: itsdangerous>=2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Flask<=2.2.3,>=1.0.4->dash==2.10) (2.2.0)\n",
      "Requirement already satisfied: click>=8.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Flask<=2.2.3,>=1.0.4->dash==2.10) (8.1.7)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from plotly>=5.0.0->dash==2.10) (8.3.0)\n",
      "Requirement already satisfied: packaging in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from plotly>=5.0.0->dash==2.10) (23.2)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from Werkzeug<=2.2.3->dash==2.10) (2.1.5)\n",
      "Requirement already satisfied: scikit-learn in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (1.4.2)\n",
      "Requirement already satisfied: numpy>=1.19.5 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from scikit-learn) (1.26.4)\n",
      "Requirement already satisfied: scipy>=1.6.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from scikit-learn) (1.12.0)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from scikit-learn) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/victoire/miniconda3/envs/ppchem/lib/python3.10/site-packages (from scikit-learn) (3.5.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install molplotly\n",
    "!pip install dash==2.10\n",
    "!pip install scikit-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# more imports\n",
    "import numpy as np\n",
    "from rdkit.Chem import AllChem, DataStructs\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Carry out PCA algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def smi_to_fp(smi):\n",
    "    '''\n",
    "    Function to convert SMILES to Morgan fingerprint\n",
    "    '''\n",
    "    fp = AllChem.GetMorganFingerprintAsBitVect(mol, radius=2, nBits=2048)\n",
    "    arr = np.zeros((0,), dtype=np.int8)\n",
    "    DataStructs.ConvertToNumpyArray(fp, arr)\n",
    "    return arr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mol' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# convert SMILES to Morgan fingerprints\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfp\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mSmiles\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43msmi_to_fp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m fps \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfp\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mtolist())\n",
      "File \u001b[0;32m~/miniconda3/envs/ppchem/lib/python3.10/site-packages/pandas/core/series.py:4764\u001b[0m, in \u001b[0;36mSeries.apply\u001b[0;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m   4629\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply\u001b[39m(\n\u001b[1;32m   4630\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4631\u001b[0m     func: AggFuncType,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4636\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   4637\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   4638\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4639\u001b[0m \u001b[38;5;124;03m    Invoke function on values of Series.\u001b[39;00m\n\u001b[1;32m   4640\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4755\u001b[0m \u001b[38;5;124;03m    dtype: float64\u001b[39;00m\n\u001b[1;32m   4756\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m   4757\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSeriesApply\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4758\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4759\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4760\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconvert_dtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4761\u001b[0m \u001b[43m        \u001b[49m\u001b[43mby_row\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby_row\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4762\u001b[0m \u001b[43m        \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4763\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m-> 4764\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/ppchem/lib/python3.10/site-packages/pandas/core/apply.py:1209\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1206\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_compat()\n\u001b[1;32m   1208\u001b[0m \u001b[38;5;66;03m# self.func is Callable\u001b[39;00m\n\u001b[0;32m-> 1209\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/ppchem/lib/python3.10/site-packages/pandas/core/apply.py:1289\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1283\u001b[0m \u001b[38;5;66;03m# row-wise access\u001b[39;00m\n\u001b[1;32m   1284\u001b[0m \u001b[38;5;66;03m# apply doesn't have a `na_action` keyword and for backward compat reasons\u001b[39;00m\n\u001b[1;32m   1285\u001b[0m \u001b[38;5;66;03m# we need to give `na_action=\"ignore\"` for categorical data.\u001b[39;00m\n\u001b[1;32m   1286\u001b[0m \u001b[38;5;66;03m# TODO: remove the `na_action=\"ignore\"` when that default has been changed in\u001b[39;00m\n\u001b[1;32m   1287\u001b[0m \u001b[38;5;66;03m#  Categorical (GH51645).\u001b[39;00m\n\u001b[1;32m   1288\u001b[0m action \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(obj\u001b[38;5;241m.\u001b[39mdtype, CategoricalDtype) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1289\u001b[0m mapped \u001b[38;5;241m=\u001b[39m \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_values\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1290\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmapper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcurried\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\n\u001b[1;32m   1291\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1293\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mapped) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(mapped[\u001b[38;5;241m0\u001b[39m], ABCSeries):\n\u001b[1;32m   1294\u001b[0m     \u001b[38;5;66;03m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[1;32m   1295\u001b[0m     \u001b[38;5;66;03m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[1;32m   1296\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m obj\u001b[38;5;241m.\u001b[39m_constructor_expanddim(\u001b[38;5;28mlist\u001b[39m(mapped), index\u001b[38;5;241m=\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/miniconda3/envs/ppchem/lib/python3.10/site-packages/pandas/core/base.py:921\u001b[0m, in \u001b[0;36mIndexOpsMixin._map_values\u001b[0;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m    918\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(arr, ExtensionArray):\n\u001b[1;32m    919\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mmap(mapper, na_action\u001b[38;5;241m=\u001b[39mna_action)\n\u001b[0;32m--> 921\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43malgorithms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_action\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/ppchem/lib/python3.10/site-packages/pandas/core/algorithms.py:1814\u001b[0m, in \u001b[0;36mmap_array\u001b[0;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m   1812\u001b[0m values \u001b[38;5;241m=\u001b[39m arr\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mobject\u001b[39m, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   1813\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m na_action \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1814\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_infer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1815\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1816\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mmap_infer_mask(\n\u001b[1;32m   1817\u001b[0m         values, mapper, mask\u001b[38;5;241m=\u001b[39misna(values)\u001b[38;5;241m.\u001b[39mview(np\u001b[38;5;241m.\u001b[39muint8), convert\u001b[38;5;241m=\u001b[39mconvert\n\u001b[1;32m   1818\u001b[0m     )\n",
      "File \u001b[0;32mlib.pyx:2926\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "Cell \u001b[0;32mIn[8], line 5\u001b[0m, in \u001b[0;36msmi_to_fp\u001b[0;34m(smi)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msmi_to_fp\u001b[39m(smi):\n\u001b[1;32m      2\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m'''\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;124;03m    Function to convert SMILES to Morgan fingerprint\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;124;03m    '''\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m     fp \u001b[38;5;241m=\u001b[39m AllChem\u001b[38;5;241m.\u001b[39mGetMorganFingerprintAsBitVect(\u001b[43mmol\u001b[49m, radius\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m, nBits\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2048\u001b[39m)\n\u001b[1;32m      6\u001b[0m     arr \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros((\u001b[38;5;241m0\u001b[39m,), dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mint8)\n\u001b[1;32m      7\u001b[0m     DataStructs\u001b[38;5;241m.\u001b[39mConvertToNumpyArray(fp, arr)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mol' is not defined"
     ]
    }
   ],
   "source": [
    "# convert SMILES to Morgan fingerprints\n",
    "df['fp'] = df['Smiles'].apply(smi_to_fp)\n",
    "fps = np.array(df['fp'].tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look up [here](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) how to use `sklearn`'s PCA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate PCA class with 2 (principal) components\n",
    "pca = # TODO code here\n",
    "# fit and transform our data (fingerprints)\n",
    "components = pca.fit_transform(fps.reshape(-1, 2048))\n",
    "# store components in dataframe\n",
    "df['PCA-1'] = # TODO code here to get first components\n",
    "df['PCA-2'] = # TODO code here to get second components"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot the first two principal components\n",
    "We already now how to plot data in a straight-forward way using `matplotlib`. While, it is very flexible we have very complex data and would like to see the influence of the molecular structure on the results. Therefore, we will look into the package[ `molplotly`](https://github.com/wjm41/molplotly/tree/main), which allows interactive plotting including much more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# plot the PCA components in a scatter plot\n",
    "fig, ax = plt.subplots()\n",
    "scatter = # TODO code here to plot PCA-1 vs PCA-2\n",
    "ax.set_xlabel('PCA-1');\n",
    "ax.set_ylabel('PCA-2');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the most basic implementation of such a PCA plot. If you like you can try to populate with more data such as colouring by drug types. In the next step we will look at how this is done with [`molplotly`](https://github.com/wjm41/molplotly/tree/main)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import molplotly\n",
    "\n",
    "# TODO: fill in the names of the columns that are needed for the plot (as strings)\n",
    "\n",
    "# instantiate the plot as done with the base package plotly\n",
    "fig_pca = px.scatter(df, # data to pull from\n",
    "                     x=, # x-axis\n",
    "                     y=, # y-axis\n",
    "                     color=, # column to color by (drug action)\n",
    "                     title='PCA of morgan fingerprints', # title of the plot\n",
    "                     labels={'color': XXXX}, # label for the color legend (same as \"color\")\n",
    "                     width=1200,\n",
    "                     height=800)\n",
    "\n",
    "# populate with molecule information\n",
    "app_pca = molplotly.add_molecules(fig=fig_pca,  # figure to populate\n",
    "                                  df=df,  # dataframe with molecule information\n",
    "                                  smiles_col=,  # column with SMILES\n",
    "                                  title_col='Name',  # column to take the name for the molecule when hovering over it\n",
    "                                  caption_cols=['USAN Definition'], # columns to take the additional captions from\n",
    "                                  color_col=, # column to color by\n",
    "                                  show_coords=False)\n",
    "\n",
    "# show the plot: don't use the same port several times as it will overwrite the previous one\n",
    "app_pca.run_server(mode='inline', port=8100, height=850)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, this plot is already much more informative. We can see the different drug groups and when hovering over the plot we can see the molecule the data point represents including the full USAN Definition and the name of the drug. We can also zoom in and out again!\n",
    "\n",
    "This plot showed colouring based on discrete groups. What happens when we want to plot a continuous variable? We will look into this in the next section by calculating the logP of all the molecules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem import Descriptors\n",
    "\n",
    "df[\"LogP\"] = # TODO code here to calculate LogP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import molplotly\n",
    "\n",
    "# TODO fill in the missing column names and color by logP this time\n",
    "\n",
    "fig_pca = px.scatter(df,\n",
    "                     x=,\n",
    "                     y=,\n",
    "                     color=,\n",
    "                     title='PCA of morgan fingerprints',\n",
    "                     labels={'color': XXXX},\n",
    "                     width=1200,\n",
    "                     height=800)\n",
    "\n",
    "app_pca = molplotly.add_molecules(fig=fig_pca,\n",
    "                                  df=df,\n",
    "                                  smiles_col=,\n",
    "                                  title_col='Name',\n",
    "                                  caption_cols=['application'],\n",
    "                                  color_col=,\n",
    "                                  show_coords=False)\n",
    "\n",
    "app_pca.run_server(mode='inline', port=8200, height=850)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This looks interesting too! However, we can see that there are some outliers that make the visulaization not so informative (and a s a side note also are not clustered very well with their group). Let's exclude those points and look at the plot again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out logP below -5 and above 5\n",
    "df_fil = df[(df['LogP'] > -5) & (df['LogP'] < 5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO repeat the same thing as above but with the filtered dataframe\n",
    "# TODO don't forget to change the port number"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this last plot the differences in logP can be more easily observed. We can see that there is some trend that the clustering coincides with the logP. This is a very interesting observation but much more analysis would have to be done to get to the depths of this. This is a very simple example of how to use PCA to visualize and analyze chemical data. There are many more ways to use PCA and many more ways to visualize the data. We will look at some more in the next section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering\n",
    "Clustering data is a way to group data points that are similar to each other. There are many different clustering algorithms, and we will look at the k-means algorithm and the Taylor-Butina algorithm. Clsutering is mostly used when we don't know the underlying data structure and need it for analysis or data processing. Another use case is evaluating how meaningful our chosen representation (in our case the Morgan fingerprint) in the given context (here drug actions) is."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### K-means clustering\n",
    "First, we will look into clustering using the k-means algorithm, which is simple and widely used. It tries to find cluster centers that are representative of certain regions of the data. The algorithm alternates between assigning data points to the nearest cluster center (based on some distance function) and updating cluster centers to be the mean of the data points that are assigned to them. We will use the [`KMeans`](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html) class from the `sklearn` package to perform k-means clustering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# even more imports\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define number of clusters based on number of drug actions\n",
    "num_clusters = # TODO What is the number of broad drug actions\n",
    "\n",
    "# TDOD instantiate KMeans object with num_clusters, a random state of 42 and n_init='auto'\n",
    "km = # TODO code here\n",
    "\n",
    "# fit data\n",
    "fp_km = df['fp'].tolist()\n",
    "# TODO code here to fit the data (fp_km)\n",
    "\n",
    "# predict cluster number on our dat\n",
    "clusters = # TODO code here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# store in dataframe\n",
    "df['cluster'] = clusters\n",
    "# convert to strings (so that it recognizes this as discrete classes)\n",
    "df['cluster'] = df['cluster'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the pca and color by cluster\n",
    "fig_pca_clusters = # TODO code here to plot PCA-1 vs PCA-2 and color by cluster\n",
    "\n",
    "app_pca_clusters = # TODO code here to add molecules to the plot\n",
    "\n",
    "app_pca_clusters.run_server(mode='inline', port=8400, height=850)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nice! If we compare to the previous plots, there certainly is some overlap with the true clusters. However, we also can see that it is far from perfect. This can have a multitude of reasons. Our drug classes contain heterogenous data, the Morgan fingerprints might not be the best representation of the data, or the clustering algorithm might not be the best choice. We will look into one of the classes next and see how diverse even one class is. Remember that we also comboned several USAN Definitions to simplify things. Thus, different molecule classes (defined by chemical motifs) were taken together."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Finding number of clusters\n",
    "In the example above, we knew how many clusters we have in our data so we could help the algorithm by already supplying it with the right number. Now, we will look into how to find the right number of clusters in an unsupervised (without true labels) way and try to cluster the subgroup of all anti-inflammatory drugs. \n",
    "\n",
    "We will use the Silhouette score to find the optimal number of clusters. The Silhouette score is a measure of how similar an object is to its own cluster (cohesion) compared to other clusters (separation). The score ranges from -1 to 1, where a high value indicates that the object is well matched to its own cluster and poorly matched to neighboring clusters. We will calculate the Silhouette score for different numbers of clusters and choose the number of clusters that gives the highest score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all the anti-inflammatory drugs\n",
    "antiinfl = df[df['application'] == 'anti-inflammatory'].copy()\n",
    "antiinfl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's cluster the data with a random guess of numbers of clusters\n",
    "num_clusters = 10\n",
    "\n",
    "# â€ ODO code here to intantiate, fit and predict the clusters using KMeans as above\n",
    "\n",
    "antiinfl['cluster'] = clusters\n",
    "# convert to strings\n",
    "antiinfl['cluster'] = antiinfl['cluster'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we need to repeat the PCA for the subselection of data\n",
    "pca = PCA(n_components=2)\n",
    "fsp_antinfl = np.array(antiinfl['fp'].tolist())\n",
    "components = pca.fit_transform(fsp_antinfl.reshape(-1, 2048))\n",
    "antiinfl['PCA-1'] = components[:, 0]\n",
    "antiinfl['PCA-2'] = components[:, 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show another interactive plot and color by newly determined cluster identity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_pca = # TODO code here to plot PCA-1 vs PCA-2 and color by cluster\n",
    "\n",
    "app_pca = # TODO code here to add molecules to the plot\n",
    "\n",
    "app_pca.run_server(mode='inline', port=8500, height=850)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks good! Let's have a more rational approach on determining the number of clusters. We will iterate over a range of 3 to 34 number of clusters and calculate the [`silhouette score`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html). We will choose the number of clusters where the silhouette score is the highest. We can also plot the silhouette score over the cluster size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_range = range(3,35)\n",
    "score_list = []\n",
    "\n",
    "for k in cluster_range:\n",
    "    # instantiate KMeans class\n",
    "    km = # TODO code here\n",
    "\n",
    "    # fit and predict km based on the fingerprints\n",
    "    cluster_labels = # TODO code here\n",
    "\n",
    "    # calculate the silhouette score (between fP and cluster_labels)\n",
    "    score = # TODO code here\n",
    "    score_list.append([k,score])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot silhouette score over number of clusters\n",
    "score_df = pd.DataFrame(score_list,columns=['k','silhouette_score'])\n",
    "fig, ax = plt.subplots()\n",
    "ax.plot(score_df['k'],score_df['silhouette_score']);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's determine the optimal cluster size programmatically\n",
    "num_clusters_opt = score_df[score_df['silhouette_score'] == score_df['silhouette_score'].max()]['k'].values[0]\n",
    "num_clusters_opt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have the optimal cluster number determined to be 4. We can rerun the k-means algorithm with the optimal number of clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "km_opt = KMeans(n_clusters=num_clusters_opt, random_state=42, n_init=\"auto\")\n",
    "clusters_opt = km_opt.fit_predict(antiinfl['fp'].tolist())\n",
    "antiinfl['cluster_opt'] = clusters_opt\n",
    "# convert to strings\n",
    "antiinfl['cluster_opt'] = antiinfl['cluster_opt'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_pca = # TODO code here to plot PCA-1 vs PCA-2 and color by \"optimal\" clusters\n",
    "\n",
    "app_pca = # TODO code here to add molecules to the plot\n",
    "\n",
    "app_pca.run_server(mode='inline', port=8600, height=850)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This looks pretty convincing. We can see that the clustering is much better than before. However, we can also see that there are still some data points that are still clustered in a weird way. This is a very common problem in clustering and can have many reasons but does not mean that anything went wrong with the clustering. As mentioned the before, the data representation has a huge influence but also the way that we visualize it is important. We will look into another clustering algorithm in the next section to see a last approach."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Taylor-Butina clustering\n",
    "Now. let's look at a different clustering algorithm, the Taylor-Butina clustering algorithm. This algorithm is based on the Tanimoto similarity between fingerprints so that all molecules in a cluster are more similar to the centroid than a pre-defined threshold but are potentially less similar to each other. This should result in small but homogenuous clusters. The steps to the algorithm are:\n",
    "1. Calculate a fingerprint (e.g. Morgan fingerprint) for each molecule in the dataset.\n",
    "2. Calculate the Tanimoto distance matrix\n",
    "3. Cluster with the Taylor-Butina algorithm: \n",
    "    1. Centroids are initiated and sorted based on number of neighbors (Tanimoto similarity above (distance below) certain threshold).\n",
    "    2. The centroid with the most neighbors gets all those assigned in its cluster.\n",
    "    3. Move to the next biggest centroid group and assign.\n",
    "    4. Repeat until no molecules with neighbors left (can have singletons - molecules without neighbors not assigned to a cluster)\n",
    "\n",
    "With this algorithm one can get molecules in one cluster that are actually more similar to another cluster's centroid but were first assigned the biggest one. Thus, the choice of threshold is an important parameter to tune."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.ML.Cluster import Butina\n",
    "from rdkit.Chem import rdMolDescriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mols = antiinfl['mol'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Calculate fingerprints for all molecules in the dataset.\n",
    "We have done this many times before. Let's to it again using a list comprehension in one line of code.\n",
    "\n",
    "As a reminder - the structure of a list comprehension is:\n",
    "```\n",
    "[function(item) for item in iterable]\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_list =  #TODO code here (FP: radius 2 and 2048 bits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check data type of the first element\n",
    "assert type(fp_list[0]) == DataStructs.cDataStructs.ExplicitBitVect"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Calculate the Tanimoto distance matrix\n",
    "\n",
    "We will use the [`DataStructs.BulkTanimotoSimilarity`](http://www.dalkescientific.com/writings/diary/archive/2020/10/02/using_rdkit_bulktanimotosimilarity.html) function from the `rdkit` package to calculate the Tanimoto similarity between all pairs of fingerprints. The Tanimoto similarity is a measure of how similar two sets are so we need to convert to distances by subtracting from 1 after.\n",
    "\n",
    "This function calculates the Tanimoto similarity between all pairs of fingerprints in the dataset. The output is a list of lists where the element at position (i, j) is the Tanimoto similarity between the fingerprints of molecule i and molecule j. Try to get the Tanimoto distance between the first molecule to all the others (including intself)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sims = # TODO code here\n",
    "sims"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the first element, as expected, gives similarity of one since it is the same molecule. The other values are between 0 and 1. The closer to 1 the more similar the molecules are. The function returns a list. We will build this list so that is a triangular distance matrix in the form of a list (so all rows are not the same length and concatenated). Write a loop to calculate the Tanimoto distance matrix where at every step you calculate the similarities between the current element and all *previous* elements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dists = []\n",
    "nfps = len(fp_list)\n",
    "for i in range(1,nfps):\n",
    "    sims = # TODO code here: similarity between i-th molecule and all previous ones\n",
    "    # convert similarity to dinstance (1-x)\n",
    "    dists.extend([1-x for x in sims])\n",
    "len(dists)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Apply the Taylor-Butina algorithm\n",
    "\n",
    "Check out the function docs [here](https://www.rdkit.org/docs/source/rdkit.ML.Cluster.Butina.html). The arguments that need to be passed are the following:\n",
    "- `distances`: The distance matrix calculated in the previous step.\n",
    "- `n`: The number of molecules in the dataset.\n",
    "- `cutoff`: The Tanimoto similarity threshold to assign molecules to a cluster.\n",
    "- `isDistData`: A boolean indicating whether the input is a distance matrix or similarity matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mol_clusters = Butina.ClusterData(\n",
    "    data = # TODO code here\n",
    "    nPts = # TODO code here\n",
    "    distThresh = # TODO code here (pick a reasonable Tanimoto cutoff for clustering)\n",
    "    isDistData = # TODO True or False?\n",
    "    )\n",
    "len(mol_clusters), type(mol_clusters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great! Now we have all the necessary parts to combine it all in one function that takes in as arguments the list of molecules and a given threshold and returns the list of cluster IDs (corresponding to the molecule list) and a the returned tuple of cluster indices sorted by cluster size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def butina_cluster(mol_list,cutoff=0.35):\n",
    "    '''\n",
    "    Carry out Tayloer-Butina algorithm.\n",
    "\n",
    "    :param mol_list: list of RDKit molecules\n",
    "    :param cutoff: Tanimoto cutoff value for clustering\n",
    "\n",
    "    :return: list of cluster IDs for each molecule, dictionary with cluster ID -> [molecule indices]\n",
    "    '''\n",
    "    # 1. get fingerprint\n",
    "    fp_list = # TODO code here to get fingerprints (radius 2, 2048 bits) from molecules in a list\n",
    "    dists = []\n",
    "    nfps = len(fp_list)\n",
    "\n",
    "    # 2. get tanimoto distance matrix\n",
    "    for i in range(1,nfps):\n",
    "        sims = # TODO get BulkTanimotoSimilarity of the i-th fingerprint with all previous ones\n",
    "        # convert similarity to dinstance (1-x)\n",
    "        dists.extend([1-x for x in sims])\n",
    "\n",
    "    # 3. apply Taylor-Butina algorithm (the rest happens under the hood)\n",
    "    mol_clusters = # TODO code here\n",
    "    cluster_id_list = [0]*nfps\n",
    "    for idx,cluster in enumerate(mol_clusters,1):\n",
    "        for member in cluster:\n",
    "            cluster_id_list[member] = idx\n",
    "\n",
    "    # get tuple sorted: cluster_id -> [molecule indices]\n",
    "    mol_clusters = sorted(mol_clusters, key=len, reverse=True)\n",
    "\n",
    "    return cluster_id_list, mol_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "antiinfl['butina_cluster'], cluster_dict = # TODO apply the function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the size of the clusters vs. their index\n",
    "fig, ax = plt.subplots(figsize=(15, 4))\n",
    "ax.set_xlabel(\"Cluster index\")\n",
    "ax.set_ylabel(\"Number of molecules\")\n",
    "ax.bar(range(1, len(cluster_dict) + 1), [len(c) for c in cluster_dict], lw=5);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We already addressed that the threshold determines how the clusters are built and is thus a parameter worth exploring. Before, we chose a random threshold (default value of 0.35) and can see that this clustering does not give good cluster sizes. In the following cell we will iterate over a range of thresholds and plot the cluster sizes sorted by size. Based on that we can select our optimal threshold. We want a smooth distribution while at the same time have as little singletons (clusters with one data point) as possible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cutoff in np.arange(0.0, 1.0, 0.1):\n",
    "    _, cluster_dict = butina_cluster(mols, cutoff=cutoff)\n",
    "    fig, ax = plt.subplots(figsize=(15, 4))\n",
    "    ax.set_title(f\"Threshold: {cutoff:3.1f}\")\n",
    "    ax.set_xlabel(\"Cluster index\")\n",
    "    ax.set_ylabel(\"Number of molecules\")\n",
    "    ax.bar(range(1, len(cluster_dict) + 1), [len(c) for c in cluster_dict], lw=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's choose the best threshold based on the criteria we defined above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mols = antiinfl['mol'].tolist()\n",
    "best_cutoff = # TODO code here\n",
    "antiinfl['butina_cluster'], cluster_dict = butina_cluster(mols, cutoff=best_cutoff)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's plot another interactive map as done before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_pca = px.scatter(antiinfl,\n",
    "                     x=\"PCA-1\",\n",
    "                     y=\"PCA-2\",\n",
    "                     color='butina_cluster',\n",
    "                     title='PCA of morgan fingerprints',\n",
    "                     labels={'color': 'butina_cluster'},\n",
    "                     width=1200,\n",
    "                     height=800)\n",
    "\n",
    "app_pca = molplotly.add_molecules(fig=fig_pca,\n",
    "                                  df=antiinfl,\n",
    "                                  smiles_col='Smiles',\n",
    "                                  title_col='Name',\n",
    "                                  caption_cols=['USAN Definition', 'cluster_opt', 'butina_cluster'],\n",
    "                                  color_col='butina_cluster',\n",
    "                                  show_coords=False)\n",
    "\n",
    "app_pca.run_server(mode='inline', port=8700, height=850)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because we have a lot of small clusters, the plot above is difficult to analyze. Let's visualize some molecules and see if the clustering makes sense to us from a chemistry perspective."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem import Draw\n",
    "print(\"Ten molecules from largest cluster:\")\n",
    "# Draw molecules\n",
    "Draw.MolsToGridImage(\n",
    "    [mols[i] for i in cluster_dict[0][:10]],\n",
    "    molsPerRow=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Ten molecules from second largest cluster:\")\n",
    "# Draw molecules\n",
    "Draw.MolsToGridImage(\n",
    "    [mols[i] for i in cluster_dict[1][:10]],\n",
    "    molsPerRow=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Ten molecules from third largest cluster:\")\n",
    "# Draw molecules\n",
    "Draw.MolsToGridImage(\n",
    "    [mols[i] for i in cluster_dict[2][:10]],\n",
    "    molsPerRow=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I would argue that the three visualized clusters certainly show similarity between the molecules. Well done! This analysis is only one aspect and depending on the objective can be handled differently. But clustering and visualizing the data will help with getting a sense of what you are working with and if your data processing for a project can makes sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
